File: benchmarks/bench_mlx.py
#!/usr/bin/env python3
"""
bench_mlx.py — Python MLX benchmark suite.

Mirrors the Rust mlx-rs benchmark (examples/benchmark.rs) so you can
compare language-level overhead on the same MLX C++ backend.

    pip install mlx
    python bench_mlx.py
"""

import time
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim

# ═══════════════════════════════════════════════════════════════════════════
# Helpers
# ═══════════════════════════════════════════════════════════════════════════

def bench(name: str, warmup: int, iters: int, fn):
    """Time a function over `iters` iterations after `warmup` warmups."""
    for _ in range(warmup):
        fn()

    start = time.perf_counter()
    for _ in range(iters):
        fn()
    elapsed = (time.perf_counter() - start) * 1000  # ms

    per_iter = elapsed / iters
    print(f"  {name:<40} {elapsed:>8.2f} ms total | {per_iter:>8.3f} ms/iter  ({iters} iters)")


# ═══════════════════════════════════════════════════════════════════════════
# 1. Raw Matmul Throughput
# ═══════════════════════════════════════════════════════════════════════════

def bench_matmul():
    print("\n══ MATMUL THROUGHPUT ══")
    for size in [256, 512, 1024, 2048, 4096]:
        a = mx.random.uniform(shape=(size, size))
        b = mx.random.uniform(shape=(size, size))

        def run():
            c = a @ b
            mx.eval(c)

        bench(f"matmul {size}x{size}", 3, 50, run)


# ═══════════════════════════════════════════════════════════════════════════
# 2. Element-wise Operations
# ═══════════════════════════════════════════════════════════════════════════

def bench_elementwise():
    print("\n══ ELEMENT-WISE OPS (1M elements) ══")
    n = 1_000_000
    a = mx.random.uniform(shape=(n,))
    b = mx.random.uniform(shape=(n,))

    bench("add",      3, 200, lambda: mx.eval(a + b))
    bench("multiply", 3, 200, lambda: mx.eval(a * b))
    bench("exp",      3, 200, lambda: mx.eval(mx.exp(a)))
    bench("sin",      3, 200, lambda: mx.eval(mx.sin(a)))
    bench("sqrt",     3, 200, lambda: mx.eval(mx.sqrt(mx.abs(a))))


# ═══════════════════════════════════════════════════════════════════════════
# 3. MLP Training Step
# ═══════════════════════════════════════════════════════════════════════════

class MLP(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super().__init__()
        self.layers = [
            nn.Linear(input_dim, hidden_dim),
            nn.Linear(hidden_dim, hidden_dim),
            nn.Linear(hidden_dim, output_dim),
        ]

    def __call__(self, x):
        for i, l in enumerate(self.layers):
            x = l(x)
            if i < len(self.layers) - 1:
                x = nn.relu(x)
        return x


def bench_mlp_training():
    print("\n══ MLP TRAINING STEP ══")

    configs = [
        (32, 128, 256, 10),
        (64, 512, 1024, 100),
        (128, 784, 2048, 10),
    ]

    for batch, input_dim, hidden, output in configs:
        model = MLP(input_dim, hidden, output)
        mx.eval(model.parameters())
        optimizer = optim.Adam(learning_rate=1e-3)

        x = mx.random.uniform(shape=(batch, input_dim))
        labels = mx.random.randint(0, output, shape=(batch,))

        def loss_fn(model, x, labels):
            logits = model(x)
            return mx.mean(nn.losses.cross_entropy(logits, labels))

        loss_and_grad = nn.value_and_grad(model, loss_fn)

        def step():
            loss, grads = loss_and_grad(model, x, labels)
            optimizer.update(model, grads)
            mx.eval(model.parameters(), optimizer.state, loss)

        label = f"MLP fwd+bwd+update B={batch} [{input_dim}->{hidden}->{hidden}->{output}]"
        bench(label, 3, 100, step)


# ═══════════════════════════════════════════════════════════════════════════
# 4. CNN Training Step
# ═══════════════════════════════════════════════════════════════════════════

class CNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(1, 16, kernel_size=3, stride=2, padding=1)
        self.conv2 = nn.Conv2d(16, 32, kernel_size=3, stride=2, padding=1)
        self.fc = nn.Linear(1568, 10)

    def __call__(self, x):
        x = nn.relu(self.conv1(x))
        x = nn.relu(self.conv2(x))
        x = x.reshape(x.shape[0], -1)
        return self.fc(x)


def bench_cnn_training():
    print("\n══ CNN TRAINING STEP (MNIST-like) ══")
    batch = 32
    model = CNN()
    mx.eval(model.parameters())
    optimizer = optim.Adam(learning_rate=1e-3)

    x = mx.random.uniform(shape=(batch, 28, 28, 1))
    labels = mx.random.randint(0, 10, shape=(batch,))

    def loss_fn(model, x, labels):
        logits = model(x)
        return mx.mean(nn.losses.cross_entropy(logits, labels))

    loss_and_grad = nn.value_and_grad(model, loss_fn)

    def step():
        loss, grads = loss_and_grad(model, x, labels)
        optimizer.update(model, grads)
        mx.eval(model.parameters(), optimizer.state, loss)

    bench("CNN fwd+bwd+update B=32 [28x28x1 -> 10]", 3, 100, step)


# ═══════════════════════════════════════════════════════════════════════════
# 5. Transformer Encoder Forward
# ═══════════════════════════════════════════════════════════════════════════

class TransformerEncoder(nn.Module):
    def __init__(self, d_model, n_heads, n_layers, d_ff, dropout=0.0):
        super().__init__()
        self.layers = [
            nn.TransformerEncoderLayer(d_model, n_heads, d_ff, dropout)
            for _ in range(n_layers)
        ]

    def __call__(self, x):
        for layer in self.layers:
            x = layer(x,mask=None)
        return x


def bench_transformer():
    print("\n══ TRANSFORMER ENCODER FORWARD ══")

    configs = [
        (64, 4, 2, 32, 16),
        (128, 4, 4, 64, 16),
        (256, 8, 6, 128, 8),
    ]

    for d_model, n_heads, n_layers, seq_len, batch in configs:
        d_ff = d_model * 4
        encoder = TransformerEncoder(d_model, n_heads, n_layers, d_ff)
        mx.eval(encoder.parameters())

        x = mx.random.uniform(shape=(batch, seq_len, d_model))

        def run():
            out = encoder(x)
            mx.eval(out)

        label = f"Encoder fwd B={batch} S={seq_len} d={d_model} h={n_heads} L={n_layers}"
        bench(label, 3, 50, run)


# ═══════════════════════════════════════════════════════════════════════════
# Main
# ═══════════════════════════════════════════════════════════════════════════

if __name__ == "__main__":
    print("╔══════════════════════════════════════════════════╗")
    print("║          Python MLX Benchmark Suite              ║")
    print("║          Device: Apple Silicon GPU               ║")
    print("╚══════════════════════════════════════════════════╝")

    mx.set_default_device(mx.gpu)

    bench_matmul()
    bench_elementwise()
    bench_mlp_training()
    bench_cnn_training()
    bench_transformer()

    print("\n══ DONE ══")

File: benchmarks/bench_mlx.rs
// examples/benchmark.rs
//
// Benchmark suite for mlx-rs on Apple Silicon.
// Run with: cargo run --release --example benchmark
//
// Compares well against the Python MLX equivalent (bench_mlx.py)
// and PyTorch MPS equivalent (bench_torch.py) included in this repo.

use mlx::{Array, Dtype, Result, transforms, Device, DeviceType};
use mlx::nn::{Module, ModuleParams, Optimizer, Sequential, Linear, ReLU, Conv2d, Flatten};
use mlx::nn::{cross_entropy, Adam};
use std::rc::Rc;
//use mlx::nn::layers::normalization::LayerNorm;
//use mlx::nn::layers::embedding::Embedding;
use mlx::nn::transformers::TransformerEncoder;
use std::cell::RefCell;
use std::time::Instant;

// ═══════════════════════════════════════════════════════════════════════════
// Benchmark Helpers
// ═══════════════════════════════════════════════════════════════════════════

/// Time a closure over `n` iterations, return (total_ms, per_iter_ms).
fn bench<F: FnMut() -> Result<()>>(name: &str, warmup: usize, iters: usize, mut f: F) {
    // Warmup — don't measure
    for _ in 0..warmup {
        f().unwrap();
    }

    let start = Instant::now();
    for _ in 0..iters {
        f().unwrap();
    }
    let elapsed = start.elapsed();

    let total_ms = elapsed.as_secs_f64() * 1000.0;
    let per_iter = total_ms / iters as f64;
    println!(
        "  {:<40} {:>8.2} ms total | {:>8.3} ms/iter  ({} iters)",
        name, total_ms, per_iter, iters
    );
}

// ═══════════════════════════════════════════════════════════════════════════
// 1. Raw Matmul Throughput
// ═══════════════════════════════════════════════════════════════════════════

fn bench_matmul(key: &Array) -> Result<()> {
    println!("\n══ MATMUL THROUGHPUT ══");

    for &size in &[256, 512, 1024, 2048, 4096] {
        let a = Array::random_uniform(
            &[size, size], -1.0, 1.0, Dtype::Float32, key,
        )?;
        let b = Array::random_uniform(
            &[size, size], -1.0, 1.0, Dtype::Float32, key,
        )?;

        bench(&format!("matmul {}x{}", size, size), 3, 50, || {
            let c = a.matmul(&b)?;
            c.eval()?;
            Ok(())
        });
    }
    Ok(())
}

// ═══════════════════════════════════════════════════════════════════════════
// 2. Element-wise Operations
// ═══════════════════════════════════════════════════════════════════════════

fn bench_elementwise(key: &Array) -> Result<()> {
    println!("\n══ ELEMENT-WISE OPS (1M elements) ══");

    let n = 1_000_000;
    let a = Array::random_uniform(&[n], -1.0, 1.0, Dtype::Float32, key)?;
    let b = Array::random_uniform(&[n], -1.0, 1.0, Dtype::Float32, key)?;

    bench("add", 3, 200, || { let c = a.add(&b)?; c.eval()?; Ok(()) });
    bench("multiply", 3, 200, || { let c = a.multiply(&b)?; c.eval()?; Ok(()) });
    bench("exp", 3, 200, || { let c = a.exp()?; c.eval()?; Ok(()) });
    bench("sin", 3, 200, || { let c = a.sin()?; c.eval()?; Ok(()) });
    bench("sqrt", 3, 200, || { let c = a.abs()?.sqrt()?; c.eval()?; Ok(()) });

    Ok(())
}


// ═══════════════════════════════════════════════════════════════════════════
// 3. MLP Training Step (JIT Compiled)
// ═══════════════════════════════════════════════════════════════════════════

fn bench_mlp_training(key: &Array) -> Result<()> {
    println!("\n══ MLP TRAINING STEP ══");

    for &(batch, input, hidden, output) in &[
        (32_usize, 128_usize, 256_usize, 10_usize),
        (64, 512, 1024, 100),
        (128, 784, 2048, 10),
    ] {
        // Use Rc to share the model into the static JIT closure
        let model = Rc::new(RefCell::new(Sequential::new(vec![
            Box::new(Linear::new(input, hidden, true, key)?),
            Box::new(ReLU::new()),
            Box::new(Linear::new(hidden, hidden, true, key)?),
            Box::new(ReLU::new()),
            Box::new(Linear::new(hidden, output, true, key)?),
        ])));

        let mut optimizer = Adam::new(1e-3, &model.borrow().parameters_owned())?;

        let x = Array::random_uniform(&[batch, input], -1.0, 1.0, Dtype::Float32, key)?;
        let class_range = Array::arange(0.0, output as f64, 1.0, Dtype::Float32)?;
        let raw_labels = Array::random_uniform(&[batch], 0.0, output as f32, Dtype::Float32, key)?
            .cast(Dtype::Int32)?;
        let targets = raw_labels.reshape(&[batch as i32, 1])?
            .equal(&class_range.reshape(&[1, output as i32])?)?
            .cast(Dtype::Float32)?;

        let label = format!(
            "MLP fwd+bwd+update B={} [{}->{}->{}->{}]",
            batch, input, hidden, hidden, output
        );

        // --- NEW: JIT Compile the Forward & Backward pass ---
        let x_c = x.clone();
        let targets_c = targets.clone();
        let model_c = Rc::clone(&model);

        // This compiles the entire computation graph into a single FFI boundary call!
        let  compiled_grad_step = mlx::compile(move |p: &[Array]| -> Result<Vec<Array>> {
        let (loss, grads) = transforms::value_and_grad(|inner_p: &[Array]| {
            let logits = {
                let mut m = model_c.borrow_mut();
                m.update_parameters(inner_p);
                m.forward(&x_c)? 
            };
            cross_entropy(&logits, &targets_c) 
        }, p)?; 

        let mut out = vec![loss];
        out.extend(grads);
        Ok(out)
    }, false)?;

        bench(&label, 3, 100, || {
            let mut params = model.borrow().parameters_owned();

            // 1. ONE FFI call executes the entire compiled graph!
            let outputs = compiled_grad_step(&params)?;
            let loss = outputs[0].clone();
            let grads = outputs[1..].to_vec();

            // 2. Optimizer step remains eager (which is fast enough on its own)
            optimizer.update(params.iter_mut().collect(), grads)?;
            model.borrow_mut().update_parameters(&params);

            // 3. Force lazy evaluation
            let mut to_eval = params;
            to_eval.push(loss);
            Array::eval_all(&to_eval)?;
            Ok(())
        });
    }
    Ok(())
}

// ═══════════════════════════════════════════════════════════════════════════
// 4. CNN Training Step (JIT Compiled)
// ═══════════════════════════════════════════════════════════════════════════

fn bench_cnn_training(key: &Array) -> Result<()> {
    println!("\n══ CNN TRAINING STEP (MNIST-like) ══");

    let batch = 32;
    // Use Rc to share the model into the static JIT closure
    let model = Rc::new(RefCell::new(Sequential::new(vec![
        Box::new(Conv2d::new(1, 16, [3, 3], [2, 2], [1, 1], [1, 1], 1, true, key)?),
        Box::new(ReLU::new()),
        Box::new(Conv2d::new(16, 32, [3, 3], [2, 2], [1, 1], [1, 1], 1, true, key)?),
        Box::new(ReLU::new()),
        Box::new(Flatten::new()),
        Box::new(Linear::new(1568, 10, true, key)?),
    ])));

    let mut optimizer = Adam::new(1e-3, &model.borrow().parameters_owned())?;

    let x = Array::random_uniform(&[batch, 28, 28, 1], -1.0, 1.0, Dtype::Float32, key)?;
    let class_range = Array::arange(0.0, 10.0, 1.0, Dtype::Float32)?;
    let raw_labels = Array::random_uniform(&[batch], 0.0, 10.0, Dtype::Float32, key)?
        .cast(Dtype::Int32)?;
    let targets = raw_labels.reshape(&[batch as i32, 1])?
        .equal(&class_range.reshape(&[1, 10])?)?
        .cast(Dtype::Float32)?;

    // --- NEW: JIT Compile the Forward & Backward pass ---
    let x_c = x.clone();
    let targets_c = targets.clone();
    let model_c = Rc::clone(&model);
    let  compiled_grad_step = mlx::compile(move |p: &[Array]| -> Result<Vec<Array>> {
        let (loss, grads) = transforms::value_and_grad(|inner_p: &[Array]| {
            let logits = {
                let mut m = model_c.borrow_mut();
                m.update_parameters(inner_p);
                m.forward(&x_c)? // Use ? instead of .unwrap()
            };
            Ok(cross_entropy(&logits, &targets_c).unwrap()) // Removed .unwrap() to return the Result directly
        }, p)?; // Use ? here as well

        let mut out = vec![loss];
        out.extend(grads);
        Ok(out)
    }, false)?;

    // let mut compiled_grad_step = mlx::compile(move |p: &[Array]| -> Result<Vec<Array>> {
    //     let (loss, grads) = transforms::value_and_grad(|inner_p: &[Array]| {
    //         let logits = {
    //             let mut m = model_c.borrow_mut();
    //             m.update_parameters(inner_p);
    //             m.forward(&x_c).unwrap()
    //         };
    //         cross_entropy(&logits, &targets_c).unwrap()
    //     }, p).unwrap();

    //     let mut out = vec![loss];
    //     out.extend(grads);
    //     Ok(out)
    // }, false)?;

    bench("CNN fwd+bwd+update B=32 [28x28x1 -> 10]", 3, 100, || {
        let mut params = model.borrow().parameters_owned();
        
        // 1. ONE FFI call executes the entire compiled graph!
        let outputs = compiled_grad_step(&params)?;
        let loss = outputs[0].clone();
        let grads = outputs[1..].to_vec();

        // 2. Optimizer update
        optimizer.update(params.iter_mut().collect(), grads)?;
        model.borrow_mut().update_parameters(&params);

        // 3. Force lazy evaluation
        let mut to_eval = params;
        to_eval.push(loss);
        Array::eval_all(&to_eval)?;
        Ok(())
    });

    Ok(())
}

// // ═══════════════════════════════════════════════════════════════════════════
// // 3. MLP Training Step
// // ═══════════════════════════════════════════════════════════════════════════

// fn bench_mlp_training(key: &Array) -> Result<()> {
//     println!("\n══ MLP TRAINING STEP ══");

//     for &(batch, input, hidden, output) in &[
//         (32_usize, 128_usize, 256_usize, 10_usize),
//         (64, 512, 1024, 100),
//         (128, 784, 2048, 10),
//     ] {
//         let model = RefCell::new(Sequential::new(vec![
//             Box::new(Linear::new(input, hidden, true, key)?),
//             Box::new(ReLU::new()),
//             Box::new(Linear::new(hidden, hidden, true, key)?),
//             Box::new(ReLU::new()),
//             Box::new(Linear::new(hidden, output, true, key)?),
//         ]));

//         let mut optimizer = Adam::new(1e-3, &model.borrow().parameters_owned())?;

//         let x = Array::random_uniform(&[batch, input], -1.0, 1.0, Dtype::Float32, key)?;
//         let class_range = Array::arange(0.0, output as f64, 1.0, Dtype::Float32)?;
//         let raw_labels = Array::random_uniform(&[batch], 0.0, output as f32, Dtype::Float32, key)?
//             .cast(Dtype::Int32)?;
//         let targets = raw_labels.reshape(&[batch as i32, 1])?
//             .equal(&class_range.reshape(&[1, output as i32])?)?
//             .cast(Dtype::Float32)?;

//         let label = format!(
//             "MLP fwd+bwd+update B={} [{}->{}->{}->{}]",
//             batch, input, hidden, hidden, output
//         );

//         bench(&label, 3, 100, || {
//             let mut params = model.borrow().parameters_owned();
//             let (loss, grads) = transforms::value_and_grad(|p: &[Array]| {
//                 let logits = {
//                     let mut m = model.borrow_mut();
//                     m.update_parameters(p);
//                     m.forward(&x)?
//                 };
//                 cross_entropy(&logits, &targets)
//             }, &params)?;
//             optimizer.update(params.iter_mut().collect(), grads)?;
//             model.borrow_mut().update_parameters(&params);
//             let mut to_eval = params;
//             to_eval.push(loss);
//             Array::eval_all(&to_eval)?;
//             Ok(())
//         });
//     }
//     Ok(())
// }

// // ═══════════════════════════════════════════════════════════════════════════
// // 4. CNN Training Step
// // ═══════════════════════════════════════════════════════════════════════════

// fn bench_cnn_training(key: &Array) -> Result<()> {
//     println!("\n══ CNN TRAINING STEP (MNIST-like) ══");

//     let batch = 32;
//     let model = RefCell::new(Sequential::new(vec![
//         Box::new(Conv2d::new(1, 16, [3, 3], [2, 2], [1, 1], [1, 1], 1, true, key)?),
//         Box::new(ReLU::new()),
//         Box::new(Conv2d::new(16, 32, [3, 3], [2, 2], [1, 1], [1, 1], 1, true, key)?),
//         Box::new(ReLU::new()),
//         Box::new(Flatten::new()),
//         Box::new(Linear::new(1568, 10, true, key)?),
//     ]));

//     let mut optimizer = Adam::new(1e-3, &model.borrow().parameters_owned())?;

//     let x = Array::random_uniform(&[batch, 28, 28, 1], -1.0, 1.0, Dtype::Float32, key)?;
//     let class_range = Array::arange(0.0, 10.0, 1.0, Dtype::Float32)?;
//     let raw_labels = Array::random_uniform(&[batch], 0.0, 10.0, Dtype::Float32, key)?
//         .cast(Dtype::Int32)?;
//     let targets = raw_labels.reshape(&[batch as i32, 1])?
//         .equal(&class_range.reshape(&[1, 10])?)?
//         .cast(Dtype::Float32)?;

//     bench("CNN fwd+bwd+update B=32 [28x28x1 -> 10]", 3, 100, || {
//         let mut params = model.borrow().parameters_owned();
//         let (loss, grads) = transforms::value_and_grad(|p: &[Array]| {
//             let logits = {
//                 let mut m = model.borrow_mut();
//                 m.update_parameters(p);
//                 m.forward(&x)?
//             };
//             cross_entropy(&logits, &targets)
//         }, &params)?;
//         optimizer.update(params.iter_mut().collect(), grads)?;
//         model.borrow_mut().update_parameters(&params);
//         let mut to_eval = params;
//         to_eval.push(loss);
//         Array::eval_all(&to_eval)?;
//         Ok(())
//     });

//     Ok(())
// }

// ═══════════════════════════════════════════════════════════════════════════
// 5. Transformer Forward Pass
// ═══════════════════════════════════════════════════════════════════════════

fn bench_transformer(key: &Array) -> Result<()> {
    println!("\n══ TRANSFORMER ENCODER FORWARD ══");

    for &(d_model, n_heads, n_layers, seq_len, batch) in &[
        (64_usize,  4_usize, 2_usize, 32_usize, 16_usize),
        (128,       4,       4,       64,       16),
        (256,       8,       6,       128,      8),
    ] {
        let d_ff = d_model * 4;
        let encoder = TransformerEncoder::new(n_layers, d_model, n_heads, d_ff, 0.0, key)?;

        let x = Array::random_uniform(
            &[batch, seq_len, d_model], -1.0, 1.0, Dtype::Float32, key,
        )?;

        let label = format!(
            "Encoder fwd B={} S={} d={} h={} L={}",
            batch, seq_len, d_model, n_heads, n_layers
        );

        bench(&label, 3, 50, || {
            let out = encoder.forward(&x)?;
            out.eval()?;
            Ok(())
        });
    }
    Ok(())
}

// ═══════════════════════════════════════════════════════════════════════════
// Main
// ═══════════════════════════════════════════════════════════════════════════

fn main() -> Result<()> {
    let gpu = Device::new(DeviceType::Gpu);
    gpu.set_default()?;

    println!("╔══════════════════════════════════════════════════╗");
    println!("║          mlx-rs Benchmark Suite                  ║");
    println!("║          Device: Apple Silicon GPU               ║");
    println!("║          Build: --release                        ║");
    println!("╚══════════════════════════════════════════════════╝");

    let key = Array::key(42)?;

    bench_matmul(&key)?;
    bench_elementwise(&key)?;
    bench_mlp_training(&key)?;
    bench_cnn_training(&key)?;
    bench_transformer(&key)?;

    println!("\n══ DONE ══");
    println!("Compare against: python bench_mlx.py && python bench_torch.py");

    Ok(())
}